web/pandas/pdeps/0008-inplace-methods-in-pandas.md

```diff
@@ -0,0 +1,397 @@
+
+To summarize the status quo of inplace behavior of methods, we have divided methods that can operate inplace or have
+an ``inplace``/``copy`` keyword into 4 groups:
+
+**Group 1: Methods that always operate inplace (no user-control with ``inplace``/``copy`` keyword) **
+
+| Method Name   |
+|:--------------|
+| ``insert``    |
+| ``pop``       |
+| ``update``    |
+| ``isetitem``* |
+
+\* Although ``isetitem`` operates on the original pandas object inplace, it will not change any existing values
+inplace (it will remove the values of the column being set, and insert new values).
+
+**Group 2: Methods that modify the underlying data of the DataFrame/Series object and can be done inplace**
+
+| Method Name     |
+|:----------------|
+| ``where``       |
+| ``fillna``      |
+| ``replace``     |
+| ``mask``        |
+| ``interpolate`` |
+| ``ffill``       |
+| ``bfill``       |
+| ``clip``        |
+
+These methods don't operate inplace by default, but can be done inplace with `inplace=True` if the dtypes are compatible
+(e.g. the values replacing the old values can be stored in the original array without an astype). All those methods leave
+the structure of the DataFrame or Series intact (shape, row/column labels), but can mutate some elements of the data of
+the DataFrame or Series.
+
+**Group 3: Methods that modify the DataFrame/Series object, but not the pre-existing values**
+
+| Method Name                 | Keyword               |
+|:----------------------------|-----------------------|
+| ``drop`` (dropping columns) | ``inplace``           |
+| ``rename``                  | ``inplace``, ``copy`` |
+| ``rename_axis``             | ``inplace``, ``copy`` |
+| ``reset_index``             | ``inplace``           |
+| ``set_index``               | ``inplace``           |
+| ``astype``                  | ``copy``              |
+| ``infer_objects``           | ``copy``              |
+| ``set_axis``                | ``copy``              |
+| ``set_flags``               | ``copy``              |
+| ``to_period``               | ``copy``              |
+| ``to_timestamp``            | ``copy``              |
+| ``tz_localize``             | ``copy``              |
+| ``tz_convert``              | ``copy``              |
+| ``Series.swaplevel``*       | ``copy``              |
+| ``concat``                  | ``copy``              |
+
+\* The `copy` keyword is only available for `Series.swaplevel` and not for `DataFrame.swaplevel`.
+
+These methods can change the structure of the DataFrame or Series, such as changing the shape by adding or removing
+columns, or changing the row/column labels (changing the index/columns attributes), but don't modify the existing
+underlying data of the object.
+
+All those methods (except for `set_flags`) make a copy of the full data by default, but can be performed inplace with
+avoiding copying all data (currently enabled with the `inplace` or `copy` keyword).
```

Comment:
I'm struggling to understand the difference between group 2 and group 3

It looks to me like the groups are:

- group 1: methods that operate inplace on the pandas object, and possibly inplace on the values;
- group 2: methods where the user can choose whether to:
  - operate inplace on the pandas object (in which, it might operate inplace on the values)
  - not operate inplace on the pandas object
- group 3: ?
- group 4: methods where users might be able to affect whether they operate inplace on the pandas object, but can never operate inplace on the values

I don't understand what you mean by "but not the pre-existing values" - don't `tz_localize` and `astype` modify values?

Reply:
> I don't understand what you mean by "but not the pre-existing values" - don't `tz_localize` and `astype` modify values?

We have "These methods can change the structure of the DataFrame or Series, such as changing the shape by adding or removing columns, or changing the row/column labels (changing the index/columns attributes), but don't modify the existing underlying data of the object.". 
That's not clear enough? Can you try to explain what aspect about that sentence is not clear?

`tz_localize` doesn't affect _data_, only the index (of course the Index also consists of data, so that is maybe the confusing part). It's `.dt.tz_localize()` that changes the data itself.

`astype` -> yeah, that's a bit an odd one that doesn't really fit in any of the groups. It can never work actually inplace like group 2 (it always replaces full columns, if a cast is done), but it can be done inplace "on the object" in the meaning of not copying the columns that are not cast. 

Maybe what we should clarify is that for this group 3 the essential point is that the current inplace=True/copy=False options can avoid a copy being made (while not actually being inplace on the array itself like group 2).

Reply:
> We really need to align on what "inplace on the object" means - perhaps a glossary could be given somewhere in the PDEP?

This is also why I suggested for group 3 that we change the name of the parameter from `inplace` or `copy` to something else, which makes it clear that it is "inplace-like", but not really "inplace"

Reply:
> We really need to align on what "inplace on the object" means - perhaps a glossary could be given somewhere in the PDEP?

Yes, I mentioned earlier that we should maybe more consistently use distinct terms like "object-inplace" and "values-inplace" (https://github.com/pandas-dev/pandas/pull/51466#discussion_r1119995038), and then we of course need to define those clearly. 

(still planning to push an update for this)

Reply:
> This is also why I suggested for group 3

@Dr-Irv You mean group 2 .. ?

Reply:
> > This is also why I suggested for group 3
> 
> @Dr-Irv You mean group 2 .. ?

Actually I mean both groups 2 and 3.  I want to get rid of `inplace` so there is no confusion.